api/envoy/extensions/transport_sockets/tls/v3/tls.proto

```diff
@@ -252,13 +246,27 @@ message CommonTlsContext {
 
     // Certificate provider for fetching validation context.
     // [#not-implemented-hide:]
-    CertificateProvider validation_context_certificate_provider = 10;
+    CertificateProvider validation_context_certificate_provider = 10
+        [deprecated = true, (envoy.annotations.deprecated_at_minor_version) = "3.0"];
 
     // Certificate provider instance for fetching validation context.
     // [#not-implemented-hide:]
-    CertificateProviderInstance validation_context_certificate_provider_instance = 12;
+    CertificateProviderInstance validation_context_certificate_provider_instance = 12
+        [deprecated = true, (envoy.annotations.deprecated_at_minor_version) = "3.0"];
   }
 
+  // Certificate provider for fetching CA certs. This will populate the
+  // :ref:`CertificateValidationContext.trusted_ca field
```

Comment:
Why not put this in CertificateValidationContext to be a sibling of trusted_ca?

Reply:
For the same reason `tls_certificate_certificate_provider` is not part of `TlsCertificate` to be a sibling of `certificate_chain` field?

Reply:
I agree with Eric, it makes a lot more sense to put it in `CertificateValidationContext`.  There it will effectively be a `oneof`, rather than being something we need to merge in later.

Sanjay, a `CertificateProvider` is an alternative to `TlsCertificate`; it provides the same information in a different way.  It would not make sense to have it inside of `TlsCertificate`.

I've gone ahead and made this change.

Reply:
> ...CertificateProvider is an alternative to TlsCertificate...

`CertificateValidationContext` does not use `TlsCertificate` but instead uses `trusted_ca` (which is of type `DataSource`)  so the field `trusted_ca` in  `CertificateValidationContext` serves the exact same purpose as `certificate_chain` in `TlsCertificate` for sending the actual certificate.

Your comment also says  "Only one of *trusted_ca*, *ca_certificate_provider*, and *ca_certificate_provider_instance*".

Reply:
`TlsCertificate` contains both the certs and the private keys.  A `CertificateProvider` provdes both, so it's an alternative to the entire `TlsCertificate` message.

Reply:
That's right, I forgot the private key part. So the `TlsCertificate` and `CertificateValidationContext` are clearly differentiated that way but a CertificateProvider is used to get a cert + key for the identity certificate (like a `TlsCertificate`) or just the trust/root certificate (like the trusted_ca of CertificateValidationContext). Makes sense